### ISSUE TITLE
Hot reloading of Config

### ISSUE THREAD

**Issue by ModProg** [k0kubun/xremap#76]
I would like to hot reload my config on changes.

I would be able to implement this change myself, but would like to know your preference on how you'd like this implemented.

I thought about implementing it similarly to `watch` through:
- add Inotify to check for `IN_MODIFY` on the config file it self, and `IN_CREATE` for the directory containing the config file as some editors don't modify but replace files on save.
- when the config is changed reload it in the event_loop.

**Comment by k0kubun** [comment]
The implementation idea looks good to me. To detect "the config is changed", you could check the mtime of the file.

Do you have any idea about an option name to enable it? I thought about some possibilities:

1. Leave `--watch` as is and introduce a new option, e.g. `--watch-config` or `--reload`
   * It's good that backward compatibility is maintained, but the downside is that it might be hard to understand which of device and config `--watch` watches if we have two different options to watch things.
1. Let `--watch` watch both devices and the config
    * It might be surprising to existing users, and technically this is a breaking change. People who want to watch only devices may not be happy.
3. Break the compatibility of `--watch` and also introduce a new option, e.g. `--watch-device`/`--watch-config` or `--watch device`/`--watch config`
   * I'm okay with this, but users need to change how they start xremap.
4. Let `--watch` optionally take an option to select a watch target "device" (default) or "config"
   * This seems useful but feels a bit magical. It might make sense if we can build it like `--watch` vs `--watch=xxx` (no ambiguity in the command line parsing like `--watch` vs `--watch xxx`), but I don't think getopts supports it. If clap supports such a behavior, I'd be happy to change the library, but I'm not sure if it does.

Could you check if we can reasonably implement it like `--watch=device`, `--watch=config`, and `--watch` (same as `--watch=device`) using clap? If that works, I'd like to go with (4) to keep backward compatibility, but if not, I prefer (3) over (1) and (2) in order to keep CLI easy to understand.

**Comment by ModProg** [comment]
> Could you check if we can reasonably implement it like --watch=device, --watch=config, and --watch (same as --watch=device) using clap? If that works, I'd like to go with (4) to keep backward compatibility, but if not, I prefer (3) over (1) and (2) in order to keep CLI easy to understand. 

I think this is what you want in clap, supporting `--watch=device`, `--watch=device,config`, `--watch=config --watch=device` and `--watch`:
```rs
use clap::{ArgEnum, Parser};

#[derive(Parser, Debug)]
struct Opts {
    #[clap(
        long,
        arg_enum,
        min_values(0),
        use_delimiter(true),
        require_equals(true),
        default_missing_value = "device"
    )]
    watch: Option<Vec<WatchTargets>>,
}

#[derive(ArgEnum, Clone, Debug)]
enum WatchTargets {
    Device,
    Config,
}

fn main() {
    dbg!(Opts::parse());
}
``` 

I'll try to implement it in the next few days.

**Comment by ModProg** [comment]
> Nice! Thank you for checking it.

I created a PR for the clap changes for now:  #77 so you can give input before I implement the rest.

**Comment by ModProg** [comment]
I noticed the watch devices is not working correctly on my pc, I'd rewrite the section a little to make it not permanently reload and locking my keyboard until I kill `xremap` I'll make a PR so you can check it in the next few days.

**Comment by k0kubun** [comment]
Thanks. I think it currently has a race condition that could miss events, which might be the cause, so something like https://github.com/k0kubun/xremap/pull/64 is needed but you also have to skip re-building everything when there's no change in order to avoid an infinite loop.

**Comment by ModProg** [comment]
I created a version with a working device watcher in #78. I changed a bit more and prepared the config reloading already.

**Comment by k0kubun** [comment]
Released your features as v0.3.0. Thank you so much!
